src/main/kotlin/cacophony/automata/Determinization.kt

```diff
@@ -0,0 +1,87 @@
+package cacophony.automata
+
+import kotlin.collections.mutableListOf
+import kotlin.collections.mutableMapOf
+import kotlin.collections.setOf
+
+public fun <StateType> determinize(nfa: NFA<StateType>): DFA<Int> =
+    determinize(
+        nfa,
+        nfa.getProductions().keys.fold(mutableSetOf()) { charset, (_, char) ->
+            charset.add(char)
+            charset
+        },
+    )
+
+private fun <StateType> determinize(
+    nfa: NFA<StateType>,
+    charset: Iterable<Char>,
+): DFA<Int> {
+    val startingState = setOf(nfa.getStartingState())
+    var list = mutableListOf(startingState)
+    var dfaProductions = mutableMapOf<Set<StateType>, MutableMap<Char, Set<StateType>>>()
+
+    while (!list.isEmpty()) {
+        val states = list.removeFirst()
+
+        dfaProductions.getOrPut(states) { mutableMapOf() }.run {
+            charset.forEach { this[it] = nfa.getSetEdge(states, it) }
+        }
+    }
+
+    val setToInt = dfaProductions.keys.mapIndexed { index, states -> states to index }.toMap()
+    val acceptingStates = dfaProductions.keys.filter { it.any { nfa.isAccepting(it) } }.map { setToInt[it]!! }.toSet()
+    val productions =
+        dfaProductions
+            .flatMap { (state, edges) -> edges.map { Pair(setToInt[state]!!, it.key) to setToInt[it.value]!! } }
+            .toMap()
+    return SimpleDFA(setToInt[startingState]!!, productions, acceptingStates)
+}
+
+private fun <StateType> NFA<StateType>.epsilonClosure(states: Iterable<StateType>): MutableSet<StateType> {
+    var queue = states.toMutableList()
```

Comment:
Isn't there some Queue/Deque?

Reply:
Switched to `ArrayDeque`, I kind of assumed `MutableList` would be a linked list